Kubernetes was relax restrictions on environment variable names.

Reopen: #447 
[Kubernetes was relax restrictions on environment variable names.](https://github.com/kubernetes/kubernetes/pull/48986)
Sample code:
```yaml
          env:
            - name: "cluster.name"
              value: "elasticsearch-cluster-1550544058281"
            - name: "bootstrap.memory_lock"
              value: "true"
            - name: "discovery.zen.ping.unicast.hosts"
              value: "elasticsearch-discovery-1550544058281"
            - name: "node.master"
              value: "false"
            - name: "node.data"
              value: "true"
            - name: "ES_JAVA_OPTS"
              value: "-Xms8192m -Xmx8192m"
```
Error message:
```
javax.validation.ConstraintViolationException: Constraint Validations: name must match "^[A-Za-z_][A-Za-z0-9_]*$" on bean: EnvVar(name=cluster.name, value=elasticsearch-cluster-1550544067656, valueFrom=null, additionalProperties={})
```

@rohanKanojia : I try to fix this problem，I modified the file of you told me, [see here](https://github.com/chenlein/kubernetes-client/commit/cee8f91aee003070ecef31a1e0a1f2fd9230a08b). However I can not generate model, when I execute `make`, some errors have occurred：
```shell
chenleis-MacBook-Pro:kubernetes-model chenlei$ make
dep ensure -v
CGO_ENABLED=0 GO15VENDOREXPERIMENT=1 go build -a ./cmd/generate/generate.go
./generate > kubernetes-model/src/main/resources/schema/kube-schema.json
./generate validation > kubernetes-model/src/main/resources/schema/validation-schema.json
mvn clean install
[INFO] Scanning for projects...
<snip>
[INFO] Final Memory: 45M/1184M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jsonschema2pojo:jsonschema2pojo-maven-plugin:0.4.23:generate (default) on project kubernetes-model: Execution default of goal org.jsonschema2pojo:jsonschema2pojo-maven-plugin:0.4.23:generate failed: Path not present: kubernetes_apimachinery_pkg_runtime_ImageRawExtension -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :kubernetes-model
make: *** [build] Error 1
```
And, my kube-schema.json are very different from yours. Just like follows：
```json
{
  "id": "http://fabric8.io/fabric8/v2/Schema#",
  "$schema": "http://json-schema.org/schema#",
  "definitions": {
    "github_com_fabric8io_kubernetes_model_vendor_github_com_openshift_api_apps_v1_CustomDeploymentStrategyParams": {
      "type": "object",
      "description": "",
      "properties": {
        "command": {
          "type": "array",
          "description": "",
          "javaOmitEmpty": true,
          "items": {
            "type": "string",
            "description": ""
          }
        },
        "environment": {
          "type": "array",
          "description": "",
          "javaOmitEmpty": true,
          "items": {
            "$ref": "#/definitions/github_com_fabric8io_kubernetes_model_vendor_k8s_io_api_core_v1_EnvVar",
            "javaType": "EnvVar"
          }
        },
        "image": {
          "type": "string",
          "description": ""
        }
      },
      "additionalProperties": true,
      "javaType": "CustomDeploymentStrategyParams",
      "javaInterfaces": [
        "io.fabric8.kubernetes.api.model.KubernetesResource"
      ]
    },
    "github_com_fabric8io_kubernetes_model_vendor_github_com_openshift_api_apps_v1_DeploymentCause": {
      "type": "object",
<snip>
```
So, I can't confirm if my changes are correct.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,10 @@
 ### CHANGELOG
 
+#### 4.1-SNAPSHOT
+  Improvements
+  
+    * Fix #1361: Relax restrictions on environment variable names
+
 #### 4.1.2
   Bugs
 

--- a/kubernetes-model/kubernetes-model-annotator/src/main/java/io/fabric8/kubernetes/annotator/KubernetesTypeAnnotator.java
+++ b/kubernetes-model/kubernetes-model-annotator/src/main/java/io/fabric8/kubernetes/annotator/KubernetesTypeAnnotator.java
@@ -38,6 +38,9 @@ import java.util.regex.Pattern;
 
 public class KubernetesTypeAnnotator extends Jackson2Annotator {
 
+    // see: https://github.com/kubernetes/kubernetes/blob/6902f3112d98eb6bd0894886ff9cd3fbd03a7f79/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L315
+    private final String envNamePattern = "[-._a-zA-Z][-._a-zA-Z0-9]*";
+
     private final String nameIsDNS952LabelPattern = "[a-z]([-a-z0-9]*[a-z0-9])?";
 
     private final String nameIsDNS1123LabelPattern = "[a-z0-9]([-a-z0-9]*[a-z0-9])?";
@@ -80,6 +83,7 @@ public class KubernetesTypeAnnotator extends Jackson2Annotator {
                     .param("value", "done");
 
             annotateMetatadataValidator(clazz);
+            envNameValidator(clazz);
         } catch (JClassAlreadyExistsException e) {
             e.printStackTrace();
         }
@@ -172,6 +176,21 @@ public class KubernetesTypeAnnotator extends Jackson2Annotator {
                             .param("regexp", "^" + getObjectNamePattern(clazz) + "$")
                             .param("max", getObjectNameMaxLength(clazz));
                     }
+                } catch (JClassAlreadyExistsException e) {
+                    e.printStackTrace();
+                }
+                return;
+            }
+        }
+    }
+
+    private void envNameValidator(JDefinedClass clazz) {
+        for (Map.Entry<String, JFieldVar> f : clazz.fields().entrySet()) {
+            if (f.getKey().equals("name") && f.getValue().type().name().equals("String") && clazz.name().equals("EnvVar")) {
+                try {
+                    JAnnotationUse annotation = f.getValue().annotate(new JCodeModel()._class("javax.validation.constraints.Pattern"));
+
+                    annotation.param("regexp", "^" + envNamePattern + "$");
                 } catch (JClassAlreadyExistsException e) {
                     e.printStackTrace();
                 }

--- a/kubernetes-model/kubernetes-model-annotator/src/main/java/io/fabric8/kubernetes/annotator/KubernetesTypeAnnotator.java
+++ b/kubernetes-model/kubernetes-model-annotator/src/main/java/io/fabric8/kubernetes/annotator/KubernetesTypeAnnotator.java
@@ -187,12 +187,10 @@ public class KubernetesTypeAnnotator extends Jackson2Annotator {
     private void envNameValidator(JDefinedClass clazz) {
         for (Map.Entry<String, JFieldVar> f : clazz.fields().entrySet()) {
             if (f.getKey().equals("name") && f.getValue().type().name().equals("String") && clazz.name().equals("EnvVar")) {
-                try {
-                    JAnnotationUse annotation = f.getValue().annotate(new JCodeModel()._class("javax.validation.constraints.Pattern"));
-
-                    annotation.param("regexp", "^" + envNamePattern + "$");
-                } catch (JClassAlreadyExistsException e) {
-                    e.printStackTrace();
+                for (JAnnotationUse annotation: f.getValue().annotations()) {
+                    if (annotation.getAnnotationClass().name().equals("Pattern")) {
+                        annotation.param("regexp", "^" + envNamePattern + "$");
+                    }
                 }
                 return;
             }

--- a/kubernetes-model/kubernetes-model/src/test/java/io/fabric8/kubernetes/api/model/DeploymentTest.java
+++ b/kubernetes-model/kubernetes-model/src/test/java/io/fabric8/kubernetes/api/model/DeploymentTest.java
@@ -129,6 +129,23 @@ public class DeploymentTest {
                 .endFieldRef()
                 .endValueFrom()
                 .endEnv()
+
+                // test env name pattern "[-._a-zA-Z][-._a-zA-Z0-9]*"
+                .addNewEnv()
+                .withName("KUBERNETES-NAME")
+                .withNewValueFrom()
+                .withNewFieldRef()
+                .endFieldRef()
+                .endValueFrom()
+                .endEnv()
+                .addNewEnv()
+                .withName("KUBERNETES.CLUSTERNAME")
+                .withNewValueFrom()
+                .withNewFieldRef()
+                .endFieldRef()
+                .endValueFrom()
+                .endEnv()
+
                 .withImage("fabric8/fabric8-maven-sample-zero-config:snapshot-171129-120114-0102")
                 .withImagePullPolicy("IfNotPresent")
                 .withName("spring-boot")
@@ -285,8 +302,10 @@ public class DeploymentTest {
         Assert.assertEquals(1, deployment.getSpec().getTemplate().getSpec().getContainers().size());
         Assert.assertArrayEquals(new Object[] {"printenv"}, deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getCommand().toArray());
         Assert.assertArrayEquals(new Object[] {"HOSTNAME", "KUBERNETES_PORT"}, deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getArgs().toArray());
-        Assert.assertEquals(1, deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().size());
+        Assert.assertEquals(3, deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().size());
         Assert.assertEquals("KUBERNETES_NAMESPACE", deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().get(0).getName());
+        Assert.assertEquals("KUBERNETES-NAME", deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().get(1).getName());
+        Assert.assertEquals("KUBERNETES.CLUSTERNAME", deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().get(2).getName());
         Assert.assertEquals("fabric8/fabric8-maven-sample-zero-config:snapshot-171129-120114-0102", deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getImage());
         Assert.assertEquals("IfNotPresent", deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getImagePullPolicy());
         Assert.assertEquals("spring-boot", deployment.getSpec().getTemplate().getSpec().getContainers().get(0).getName());
@@ -451,6 +470,25 @@ public class DeploymentTest {
                 .endFieldRef()
                 .endValueFrom()
                 .endEnv()
+
+                // test env name pattern "[-._a-zA-Z][-._a-zA-Z0-9]*"
+                .addNewEnv()
+                .withName("KUBERNETES-NAME")
+                .withNewValueFrom()
+                .withNewFieldRef()
+                .withFieldPath("metadata.name")
+                .endFieldRef()
+                .endValueFrom()
+                .endEnv()
+                .addNewEnv()
+                .withName("KUBERNETES.CLUSTERNAME")
+                .withNewValueFrom()
+                .withNewFieldRef()
+                .withFieldPath("metadata.clusterName")
+                .endFieldRef()
+                .endValueFrom()
+                .endEnv()
+
                 .withImage("fabric8/fabric8-maven-sample-zero-config:snapshot-171129-120114-0102")
                 .withImagePullPolicy("IfNotPresent")
                 .withName("spring-boot")
@@ -616,8 +654,10 @@ public class DeploymentTest {
         Assert.assertEquals(1, deploymentConfig.getSpec().getTemplate().getSpec().getContainers().size());
         Assert.assertArrayEquals(new Object[] {"printenv"}, deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getCommand().toArray());
         Assert.assertArrayEquals(new Object[] {"HOSTNAME", "KUBERNETES_PORT"}, deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getArgs().toArray());
-        Assert.assertEquals(1, deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().size());
+        Assert.assertEquals(3, deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().size());
         Assert.assertEquals("KUBERNETES_NAMESPACE", deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().get(0).getName());
+        Assert.assertEquals("KUBERNETES-NAME", deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().get(1).getName());
+        Assert.assertEquals("KUBERNETES.CLUSTERNAME", deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().get(2).getName());
         Assert.assertEquals("fabric8/fabric8-maven-sample-zero-config:snapshot-171129-120114-0102", deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getImage());
         Assert.assertEquals("IfNotPresent", deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getImagePullPolicy());
         Assert.assertEquals("spring-boot", deploymentConfig.getSpec().getTemplate().getSpec().getContainers().get(0).getName());

--- a/kubernetes-model/kubernetes-model/src/test/resources/valid-deployment.json
+++ b/kubernetes-model/kubernetes-model/src/test/resources/valid-deployment.json
@@ -94,6 +94,22 @@
                     "fieldPath": "metadata.namespace"
                   }
                 }
+              },
+              {
+                "name": "KUBERNETES-NAME",
+                "valueFrom": {
+                  "fieldRef": {
+                    "fieldPath": "metadata.name"
+                  }
+                }
+              },
+              {
+                "name": "KUBERNETES.CLUSTERNAME",
+                "valueFrom": {
+                  "fieldRef": {
+                    "fieldPath": "metadata.clusterName"
+                  }
+                }
               }
             ],
             "image": "fabric8/fabric8-maven-sample-zero-config:snapshot-171129-120114-0102",

